### DISCUSSION TITLE
Background color, borders and extandable columns

### DISCUSSION THREAD

**Discussion by leomayer** [lvjr/tabularray#118] [Q&A]
I'm completely new to this package. The major reason why I'm thinking about switching is that I have problems with the current table environment to achieve something like the attached screen shot. 

The individual properties could be set easily. But with background color, borders (preferred colored) and extandable width to stretch over a complete linewidth I come to my limits. 

How would it look like in tabularray? 

![Bildschirmfoto von 2021-09-26 20-32-36](https://user-images.githubusercontent.com/2641128/134819842-a115d27b-ad79-419a-9476-583f9991dbcd.png)

I tried with [latex-tables.com](https://www.latex-tables.com/v1/index.html#) and got the following code:

```
\begin{table}
	\centering
	\begin{tabular}{l!{\color{red}\vrule}l!{\color[rgb]{0.502,0,0}\vrule}l!{\color{red}\vrule}l!{\color[rgb]{0.502,0,0}\vrule}l!{\color{red}\vrule}}
		& \multicolumn{2}{c!{\color{red}\vrule}}{{\cellcolor[rgb]{0.827,0.937,0.953}}2019}                       & \multicolumn{2}{c!{\color{red}\vrule}}{{\cellcolor[rgb]{0.827,0.937,0.953}}2020}                        \\
		& \multicolumn{1}{c}{{\cellcolor[rgb]{0.827,0.937,0.953}}win} & {\cellcolor[rgb]{0.827,0.937,0.953}}loss & \multicolumn{1}{c}{{\cellcolor[rgb]{0.827,0.937,0.953}}win} & {\cellcolor[rgb]{0.827,0.937,0.953}}loss  \\ 
		\hline
		{\cellcolor[rgb]{1,0.62,0.62}}Aggressive   & 10\%                                                        & 15\%                                     & 21\%                                                        & 25\%                                      \\ 
		\hline
		{\cellcolor[rgb]{1,0.62,0.62}}Moderate     & 7.5\%                                                       & 11\%                                     & 11.5\%                                                      & 23\%                                      \\ 
		\hline
		{\cellcolor[rgb]{1,0.62,0.62}}Conservative & 8.5\%                                                       & 12\%                                     & 8.24\%                                                      & 12\%                                      \\
		\hline
	\end{tabular}
\end{table}
```

Questions (as well with the generated code):

- I have no clue how to center all lines in the second line
- Right align doesn't seem too easy
- If I have a longer table - with more lines in it (could be duplicated as well) - how does this behave with page break? Especially when I want to maintain the header for the pages
- Is the caption under the table possible (instead of above)? Is this a package feature or do I need the `caption` package?
- Using the extenable columns: I tried it with `\textwidth` but somehow the lines AND the size got completely messed up. Would be good to know if this package would support it
- Do you use the syntax of `@{}` as well ?
- I didn't include `hlines` with colors (and/or) thicker. I guess they shouldn't be a big problem. (e.g as end of the header). Nevertheless lines raise the question which one to use? Especially when looking on the `booktabs` package which introduces additional spacing which on the long hand don't work together with colored cell (they have always an additional white space). Would be great if this couldwork here as well (the additional space after rules)

**Comment by lvjr** [comment]
Please read the manual of `tabularray` package first before reading this answer:
http://mirrors.ctan.org/macros/latex/contrib/tabularray/tabularray.pdf

```latex
\documentclass{article}

\usepackage{caption}
\usepackage{xcolor}

\usepackage{tabularray}
\UseTblrLibrary{booktabs}

\begin{document}

\section{Some Short Tables}

\begin{table}[!htpb]
\centering
\begin{tblr}{
  colspec = {lrrrr},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
}
\toprule
              & 2019  &      & 2020   &      \\
\cmidrule[lr]{2-3}
\cmidrule[lr]{4-5}
              & win   & loss & win    & loss \\ 
\midrule
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\bottomrule
\end{tblr}
\caption{Table Caption}
\end{table}

\begin{table}[!htpb]
\centering
\begin{tblr}{
  colspec = {X[l,2]X[r]X[r]X[r]X[r]},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
}
\toprule
              & 2019  &      & 2020   &      \\
\cmidrule[lr]{2-3}
\cmidrule[lr]{4-5}
              & win   & loss & win    & loss \\ 
\midrule
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\bottomrule
\end{tblr}
\caption{Table Caption}
\end{table}

\begin{table}[!htpb]
\centering
\caption{Table Caption}
\begin{tblr}{
  colspec = {|[red3,1pt]l|[red3,1pt]r|r|[red3,1pt]r|r|[red3,1pt]},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
}
\hline[1pt]
              & 2019  &      & 2020   &      \\
\hline
              & win   & loss & win    & loss \\ 
\hline[0.7pt]
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\hline[1pt]
\end{tblr}
\end{table}

\begin{table}[!htpb]
\centering
\caption{Table Caption}
\begin{tblr}{
  colspec = {|[red3,1pt]X[l,2]|[red3,1pt]X[r]|X[r]|[red3,1pt]X[r]|X[r]|[red3,1pt]},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
}
\hline[1pt,blue3]
              & 2019  &      & 2020   &      \\
\hline
              & win   & loss & win    & loss \\ 
\hline[0.7pt]
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\hline[1pt,blue3]
\end{tblr}
\end{table}

\newpage

\section{Some Long Tables}

\begin{longtblr}[
  caption = {Table Caption},
]{
  colspec = {lrrrr},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
  rowhead = 2, % repeat first two rows
}
\toprule
              & 2019  &      & 2020   &      \\
\cmidrule[lr]{2-3}
\cmidrule[lr]{4-5}
              & win   & loss & win    & loss \\ 
\midrule
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\bottomrule
\end{longtblr}

\begin{longtblr}[
  caption = {Table Caption},
]{
  colspec = {X[l,2]X[r]X[r]X[r]X[r]},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
  rowhead = 2, % repeat first two rows
}
\toprule
              & 2019  &      & 2020   &      \\
\cmidrule[lr]{2-3}
\cmidrule[lr]{4-5}
              & win   & loss & win    & loss \\ 
\midrule
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\bottomrule
\end{longtblr}

\begin{longtblr}[
  caption = {Table Caption},
]{
  colspec = {|[red3,1pt]l|[red3,1pt]r|r|[red3,1pt]r|r|[red3,1pt]},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
  rowhead = 2, % repeat first two rows
}
\hline[1pt]
              & 2019  &      & 2020   &      \\
\hline
              & win   & loss & win    & loss \\ 
\hline[0.7pt]
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\hline[1pt]
\end{longtblr}

\begin{longtblr}[
  caption = {Table Caption},
]{
  colspec = {|[red3,1pt]X[l,2]|[red3,1pt]X[r]|X[r]|[red3,1pt]X[r]|X[r]|[red3,1pt]},
  row{1,2} = {c,azure8,font=\bfseries},
  column{1} = {purple8,font=\bfseries},
  cell{1}{1} = {r=2}{m,white}, %multirow
  cell{1}{2,4} = {c=2}{c}, % multicolumn,
  rowhead = 2, % repeat first two rows
}
\hline[1pt,blue3]
              & 2019  &      & 2020   &      \\
\hline
              & win   & loss & win    & loss \\ 
\hline[0.7pt]
 Aggressive   & 10\%  & 15\% &   21\% & 25\% \\ 
\hline
 Moderate     & 7.5\% & 11\% & 11.5\% & 23\% \\ 
\hline
 Conservative & 8.5\% & 12\% & 8.24\% & 12\% \\
\hline[1pt,blue3]
\end{longtblr}

\end{document} 
```

**Comment by leomayer** [comment]
Thank you very much. The samples look good and I need to investigate them further. Just few questions:

1. The package `ninecolor` is required as soon as I load `xcolor`. But this seems quite strange when I don't use the colors from this package. Background: I need to define my own colors and the chances to use this package are quite slim. Wouldn't it be better to load this package via package options? Or separately i.e. `\UseTblrLibrary{ninecolors}`
2. The `line` raised the question about centering. More or less the distance to the top and bottom. Perhaps I miss something in the typography but I would have guessed that the distance from the screenshot should be same
![Bildschirmfoto von 2021-09-27 10-38-29](https://user-images.githubusercontent.com/2641128/134876155-0ff53cd0-30e1-4e58-a236-062457190fcb.png)
3. Is there a summary of the meanings for the `colspec`? I found e. g. how other column types are defined as `Q` type but I don't find a taxative summary what `l`, `c`.... nor what `Q` means. Well some of the types could easily be guessed but reading `X[l,2]` raises the question what the two parameters mean? As well for `|[red3,1pt]X[r]`. I guess the first parmeter is the color the second the thickness. I'm not sure if its only one parameter which one is taken and which won't left out. But for `l, 2` I have no real clue. In chapter 2.5.2 there is some wording but I read it somewhere else what `h` stands for and `t`. From the definitions they must be different - is it so? Where is it described what the `ht` parameters means in the `Q` definition? Are the types somewhere together described? If so, which chapter? 

For the other details I need to check. Again thx very much :-)

**Comment by lvjr** [comment]
(1) It is hard for a normal user to choose matching colors. Your original table looks bad because background colors (red and blue) in your code have different luminance values. And I have seen plenty of terrible colors in LaTeX tables. Therefore I strongly suggest every user to use predefined colors in [`ninecolors`](https://ctan.org/pkg/ninecolors) package unless he/she is a master of colors. 

(2) LaTeX tables is mainly designed for normal text which has both [ascender](https://en.wikipedia.org/wiki/Ascender_(typography)) and [descender](https://en.wikipedia.org/wiki/Descender) . It looks good if you replace `10%` with the word `Typography`.

(3a) In `colspec`, the Nth `Q` stands for the Nth column. If you want to style the column, you add some specifications of `key=value` in the optional argument of it. All available specifications for columns are listed in "Table 2.7: Keys for Columns" of the manual. For example, you can write `Q[halign=c,bg=blue8]`. Also below that table, it is stated that some key names, such as `halign` and `bg`, can be omitted. Therefore, we can write this example in a simplified form as `Q[c,blue8]`.

(3b) In `colspec`, `|` before or after some `Q` column stands for a `vline` (vertical rule) at that position.  If you want to style the `vline`, you add some specifications of `key=value` in the optional argument of it. All available specifications for vlines are listed in "Table 2.3: Keys for Vlines" of the manual. For example, you can write `|[fg=red3,wd=1pt]`. Also below that table, it is stated that some key names, such as `fg` and `wd`, can be omitted. Therefore, we can write this example in a simplified form as `|[red3,1pt]`. 

(3c) In Section 1.7 of the manual, it is stated that `X` column is defined as `Q` column with `co=1` (`co` is the coefficient), Therefore `X[r]` is the same as `Q[halign=r,co=1]` or `Q[r,1]` since `halign` and `co` keys can be omitted, and `X[l,2]` is the same as `Q[halign=l,co=2]` or `Q[l,2]` since `halign` and `co` keys can be omitted. These coefficients are used for distributing extra horizontal space to each columns. For example, in table 2 of the above code, we have `colspec = {X[l,2]X[r]X[r]X[r]X[r]}`, therefore we distribute extra space by ratio 2:1:1:1:1 to columns.

(3d) Values `h`, `t`, `m`, `b`, and `f` are for different types of vertical alignment (head, top, middle, bottom, foot). Please look carefully the example in Section 1.3 to find out the differences among them. Note that `Q[h,4cm]` is a simplified form of `Q[valign=h,wd=4cm]`. At last, the `ht` key is described in `Table 2.6: Keys for Rows`.

**Reply by leomayer** [reply]
(1) I agree with you to some extend. But using it as a mandatory package is little bit strange. Additionally I want remark, that `row{1,2} = {c,azure2,font=\bfseries},` would require to change the font color which isn't be done. Which raises the question how to set the matching `fg`? (`{c,azure2, fg=azure2}`) won't be the solution.
(2) Agreed. Thx for the hint with `Typography`
(3) Hmmm... First of all thank you for your detailed response. Which raises for me the question _Why it is not included in your manual?_ I'm thinking about it and I would guess that the information is quite cluttered.. The samples from Section 1 are good but I would arrange them differently. I get an idea about how to use the package but would expect to have some kind of 'drill down' - from the simple sample to more extended ones. I read `Column Types` and would expect either a sample and/or a hint what to use next. From a beginner's perspective I would ask _Which are the `Column Types`?_ But the section doesn't provide details, neither does `Row Types`. I guess they should be somehow specified - most likely in Section 2.3 (due to your feedback).

The remark `In most cases, you can omit the underlined key names and write only their values.` doesn't mean anything for me. First of all an underlined text doesn't catch my attention - I would expect some kind of footnote, like *. Next I would expect a specific sample, e.g. `Q[halign=r,co=1]` could be written due to their default values as `Q[r,1].

Just as a short (?) feedback. Many thanx for your excellent package.

**Reply by lvjr** [reply]
In `row{1,2} = {c,azure2,font=\bfseries}`, `azure2` is the same as `bg=azure2`. And you will not want to write `row{1,2} = {c,azure2,fg=azure2}`, which is the same as `row{1,2} = {c,bg=azure2,fg=azure2}`.

**Reply by leomayer** [reply]
Sorry, I'm confused. Using `azure2` (in `row{1,2} = {c,azure2,font=\bfseries}`)has the following result
![Bildschirmfoto von 2021-09-29 08-27-48](https://user-images.githubusercontent.com/2641128/135214785-078e2f92-834c-406b-8bad-050525d5ff60.png)
If I follow your argument about contrast I would say the solution is incorrect. So, how to set it properly?

**Reply by lvjr** [reply]
From the [manual](http://mirrors.ctan.org/macros/latex/contrib/ninecolors/ninecolors.pdf) of ninecolors package, azure2 is of level 2, and the foreground color black is of level 0, therefore they don't match and the result is bad. You need to select a foreground color of level at least 7, such as blue7, purple8, gray9, or white (which is of level 10).

**Reply by Dojh342** [reply]
Sorry to necro an old discussion:

1. Is it possible to use custom colors and/or NOT use ninecolors? I have specific brand colors I wish to use and have no need for ninecolors which seems to cause issues with xcolor.

2. Is it possible to modify the vertical alignment as was being discussed in the previous question?

Thank you for producing a wonderful table package!

**Reply by muzimuzhi** [reply]
@Dojh342 

> 1. Is it possible to use custom colors and/or NOT use `ninecolors`?

Of course yes. `ninecolors` is nothing more than a set of carefully selected colors with proper WCAG color contrast, [defined by `\definecolor` and probably also by `\color_set:nnn`](https://github.com/lvjr/ninecolors/blob/e9d5e9fa29dc2e0ac752f532005a17a725c2e637/ninecolors.sty#L382-L391).

> 2. Is it possible to modify the vertical alignment as was being discussed in the previous question?

Try out `stretch=0, rows={ht=\baselineskip}`, excerpted from the second example in package document (2023A (2023-03-01)), sec. 3.1.2 "Minimal Strut for Cell Text".

**Reply by Dojh342** [reply]
@muzimuzhi  - thank you for the speedy reply and your help!

So I have an issue where my custom colours seem to be washed out:
![image](https://github.com/lvjr/tabularray/assets/12612308/4123ab27-4a5a-4e94-a17e-10736154fdc9)

```
\documentclass[a4paper,11pt]{report}
\usepackage{xcolor}
\usepackage{tabularray}
\definecolor{nicebluehsb}{HSB}{215,62,67}
\definecolor{nicebluergb}{RGB}{65,109,171}
\begin{document}
\begin{longtblr}[
	label={table:test},
	caption={},
	]{
		width=0.9\textwidth,
		colspec={X[1,c,m]X[2,c,m]X[1,l,m]},
		rowhead=1,
		row{1}={bg=nicebluergb,fg=white,1.5em,font=\large\bfseries},
		hline{1,Z}={0.5pt},
		vlines={1pt,solid},
		stretch=0,
		rows={ht=\baselineskip},
		row{even}={bg=white},
		row{odd}={bg=nicebluergb!15},
	}
A1 & B1 & C1\\
A2 & B2 & C2\\
A3 & B3 & C3\\
\end{longtblr}

\fcolorbox{black}{nicebluergb}{\color{white}For comparison}

\end{document}
```

Any idea what's going on?

On 2. thank you! Apparently I completely failed to spot that example! Thank you again.

**Reply by muzimuzhi** [reply]
Set `row{1}={...}` _after_ `row{odd}={...}`, so the special setting for the first row won't be overwritten. This is also how the `rowhead` and `rowfoot` example in package documentation does, see package doc (2023A (2023-03-01)), page 38.

(a bit off-topic, compared to the title of current discussion...)

**Reply by Dojh342** [reply]
! That was it! I feel very silly. Thank you so much and apologies again for hijacking this thread - I think of it as a fork of the thread :D
